### ISSUE TITLE
How to change batch size when resuming training?

### ISSUE THREAD

**Issue by Les1ie** [ultralytics/ultralytics#4623]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I would like to change the batch size when I resume training.
For example, I am using 4 gpu's for the first training and the batch size is 4\*32, after a certain time of training I have 5 available gpu's, I want to reset the training while resuming the training and resize the batch size to 5\*32, how can I do that.
As of now, even if I change the "device" and "batch" in the training code and set the
"resume" to True in the training code, it will split the 4\*32 batch to 5 gpus without using the new batch size.

-----
我想在恢复训练时更改批量大小。
例如，我首次训练时使用4个gpu，batch size 为4\*32，训练了一定时间后我拥有了5个可用的gpu，我希望恢复训练的同时将batch size调整为5\*32，请问我该如何操作。
目前来看，我即使在训练的代码中修改了"device"和"batch"并将"resume"设置为True，但它会将 4\*32 的batch均分到5个gpus上，而不会使用新的batch size。

### Additional

The example code is as follows.

code for initial training:
```python
device = [0, 1, 2, 3]
model = YOLO('yolov8n.pt')
model.train(data='/data/my_dataset.yaml',
            device=device,
            batch=32* len(device), resume=False,
            )
```

code for resume training:
```python
device = [0, 1, 2, 3, 4]
model = YOLO('weights/last.pt')
model.train(data='/data/my_dataset.yaml',
            device=device,
            batch=32* len(device), resume=True,
            )
```

**Comment by glenn-jocher** [comment]
@Les1ie in Ultralytics YOLOv8, the resume functionality uses values supplied in previous training sessions to ensure continuity in the training process. Consequently, when you resume the training with a new batch size or on additional GPUs, it may still use the batch size information preserved from the previous sessions rather than the new values.

If you wish to change the batch size during a resume operation - for instance, when scaling from 4 GPUs to 5 GPUs as in your case - an effective way is to initiate a new training session with your desired batch size.

Remember that the weight files are not tied to a given batch size, so you can set these up in the new session with the weights from your previously interrupted session. This way, you're essentially resuming your training but with a new batch size and potentially more GPU power.

Your configuration appears to be correct, just keep in mind to initiate a new training session instead of attempting to resume the previous one with reconfigured settings. 

Good luck with your further training!

**Comment by Les1ie** [comment]
> @Les1ie in Ultralytics YOLOv8, the resume functionality uses values supplied in previous training sessions to ensure continuity in the training process. Consequently, when you resume the training with a new batch size or on additional GPUs, it may still use the batch size information preserved from the previous sessions rather than the new values.
> 
> If you wish to change the batch size during a resume operation - for instance, when scaling from 4 GPUs to 5 GPUs as in your case - an effective way is to initiate a new training session with your desired batch size.
> 
> Remember that the weight files are not tied to a given batch size, so you can set these up in the new session with the weights from your previously interrupted session. This way, you're essentially resuming your training but with a new batch size and potentially more GPU power.
> 
> Your configuration appears to be correct, just keep in mind to initiate a new training session instead of attempting to resume the previous one with reconfigured settings.
> 
> Good luck with your further training!

Thank you very much for your reply.😁

The reason why I tried to resume the training instead of opening a new training session is because I am worried that I will lose the information related to the training process such as `cos_lr` and `patience` from the previous training, which may lead to differences in the results, is there a way to address this concern?

**Comment by glenn-jocher** [comment]
@Les1ie yes, you're correct that certain hyperparameters and scheduler states from the previous training session, like cosine learning rate schedule (`cos_lr`) and early stopping patience (`patience`), are indeed saved and utilized when resuming training. Starting a new training session will initialize these states anew, potentially leading to slight differences in training dynamics.

If you want to retain these states while changing the batch size, one potential workaround is to manually modify the saved training states in the checkpoint file. You could load the checkpoint file (which is a pytorch .pt file), change the batch size in the 'hyp' dictionary, and then save it back. 

Please be cautious with this though and make sure the modified batch size is appropriate for your new training configuration. 

I hope this helps! Let us know if you have any more questions.

**Comment by Les1ie** [comment]
Thank you for your reply, it is helpful!

**Comment by glenn-jocher** [comment]
@Les1ie you're welcome! I'm glad to hear that you found the information helpful. Don't hesitate to reach out if you have any more questions or issues. Happy training with YOLOv8!
